packages/geth-rpc-gateway/main.go

```diff
@@ -0,0 +1,208 @@
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/gorilla/websocket"
+)
+
+type JSONRPCRequest struct {
+	Method string `json:"method"`
+}
+
+var upgrader = websocket.Upgrader{
+	CheckOrigin: func(r *http.Request) bool { return true },
+}
+
+var methodsUsingPrimary map[string]bool
+
+func main() {
+	// Load the target URLs from environment variables
+	primaryURLStr := os.Getenv("TARGET_URL_PRIMARY")
+	secondaryURLStr := os.Getenv("TARGET_URL_SECONDARY")
+	webSocketURLStr := os.Getenv("WEBSOCKET_TARGET_URL")
+	webSocketSecondaryURLStr := os.Getenv("WEBSOCKET_TARGET_URL_SECONDARY")
+
+	if primaryURLStr == "" || secondaryURLStr == "" || webSocketURLStr == "" || webSocketSecondaryURLStr == "" {
+		log.Fatal("TARGET_URL_PRIMARY, TARGET_URL_SECONDARY, WEBSOCKET_TARGET_URL, and WEBSOCKET_TARGET_URL_SECONDARY environment variables must be set")
+	}
+
+	primaryURL, err := url.Parse(primaryURLStr)
+	if err != nil {
+		log.Fatalf("Failed to parse primary target URL: %v", err)
+	}
+	secondaryURL, err := url.Parse(secondaryURLStr)
+	if err != nil {
+		log.Fatalf("Failed to parse secondary target URL: %v", err)
+	}
+	webSocketURL, err := url.Parse(webSocketURLStr)
+	if err != nil {
+		log.Fatalf("Failed to parse WebSocket target URL: %v", err)
+	}
+	webSocketSecondaryURL, err := url.Parse(webSocketSecondaryURLStr)
+	if err != nil {
+		log.Fatalf("Failed to parse WebSocket secondary target URL: %v", err)
+	}
+
+	methodsUsingPrimary = parsePrimaryMethods(os.Getenv("PRIMARY_METHODS"))
+	enableDebugEndpoints := os.Getenv("ENABLE_DEBUG_ENDPOINTS") == "true"
+
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		log.Printf("/healthz Received request: Method=%s, Path=%s", r.Method, r.URL.Path)
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("OK"))
+	})
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if strings.ToLower(r.Header.Get("Upgrade")) == "websocket" {
+			// Determine the WebSocket target URL based on the method
+			var wsTargetURL *url.URL
+			method := r.URL.Query().Get("method")
+			if shouldUsePrimaryURL(method) {
+				wsTargetURL = webSocketURL
+				log.Printf("WebSocket request hitting WEBSOCKET_TARGET_URL with method: %s", method)
+			} else {
+				wsTargetURL = webSocketSecondaryURL
+				log.Printf("WebSocket request hitting WEBSOCKET_TARGET_URL_SECONDARY with method: %s", method)
+			}
+			handleWebSocket(w, r, wsTargetURL)
+			return
+		}
+
+		bodyBytes, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "Failed to read request body", http.StatusBadRequest)
+			return
+		}
+		defer r.Body.Close()
+
+		if len(bodyBytes) == 0 {
+			w.Write([]byte("OK"))
+			return
+		}
+
+		var jsonReq JSONRPCRequest
+		if err := json.Unmarshal(bodyBytes, &jsonReq); err != nil {
+			http.Error(w, "Invalid JSON in request body", http.StatusBadRequest)
+			return
+		}
+
+		// Determine the HTTP target URL based on the method
+		var targetURL *url.URL
+		if shouldUsePrimaryURL(jsonReq.Method) {
```

Comment:
And maybe later on we can add this feature: `if a request can't be responded by primary url, we try secondary url?`

Reply:
Good idea.